aten/src/ATen/native/cuda/TensorCompare.cu

```diff
@@ -24,5 +26,95 @@ Tensor _s_where_cuda(const Tensor& condition, const Tensor& self, const Tensor&
   return ret;
 }
 
+template<template<typename T> class Comparator, typename scalar>
+struct CmpOpCUDA {
+  static void apply(Tensor& ret, const Tensor& self, Scalar other) {
+    auto other_val = other.to<scalar>();
+    cuda::CUDA_tensor_apply2<uint8_t, scalar>(ret, self,
+        [other_val] __device__ (uint8_t& ret_val, const scalar& self_val) {
+        ret_val = Comparator<scalar>()(self_val, other_val);
+      }
+    );
+  }
+};
+
+template<typename scalar>
+using LeOpCUDA = CmpOpCUDA<std::less_equal, scalar>;
+template<typename scalar>
+using GeOpCUDA = CmpOpCUDA<std::greater_equal, scalar>;
+template<typename scalar>
+using EqOpCUDA = CmpOpCUDA<std::equal_to, scalar>;
+template<typename scalar>
+using NeOpCUDA = CmpOpCUDA<std::not_equal_to, scalar>;
+
+template<template<typename T> class Comparator, typename scalar>
+struct CmpOpDoubleCUDA {
+  static void apply(Tensor& result, const Tensor& self, Scalar other) {
+    auto other_val = other.to<double>();
+    cuda::CUDA_tensor_apply2<uint8_t, scalar>(result, self,
+        [other_val] __device__ (uint8_t& result_val, const scalar& self_val) {
+        result_val = Comparator<double>()(ScalarConvert<scalar, double>::to(self_val), other_val);
```

Comment:
This will be very slow on CUDA (because of the `double` math).

Reply:
I wonder if we shouldn't use instead [`THCNumerics` for that](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCNumerics.cuh#L21-L26)

Reply:
Not sure how `THCNumerics` would help. `Comparator` is a struct with a trivial constructor and a call operator, so it should compile to equivalent code.

Reply:
hmm I guess it might fail with half. We could add explicit overloads for half in those std comparators...

Reply:
Should we stop using already implemented `THC` functions in favor of more modern C++ present in ATen? I'm not sure about how the transition should happen, and if we should slowly stop relying on `TH`/`THC` for new features.

Reply:
I'm no ATen expert. @gchanan or @colesbury would know better

Reply:
If I just use half instead of double here, would that be okay?